WreckEngine/UI.cpp

```diff
@@ -0,0 +1,402 @@
+                {
+                    GL_CHECK(glBindTexture(GL_TEXTURE_2D, (GLuint)(intptr_t)pcmd->TextureId));
+                    GL_CHECK(glScissor((int)pcmd->ClipRect.x, (int)(fb_height - pcmd->ClipRect.w), (int)(pcmd->ClipRect.z - pcmd->ClipRect.x), (int)(pcmd->ClipRect.w - pcmd->ClipRect.y)));
+                    GL_CHECK(glDrawElements(GL_TRIANGLES, (GLsizei)pcmd->ElemCount, sizeof(ImDrawIdx) == 2 ? GL_UNSIGNED_SHORT : GL_UNSIGNED_INT, idx_buffer_offset));
+                }
+                idx_buffer_offset += pcmd->ElemCount;
+            }
+        }
+    }
+
+    /**
+     * \brief Creates the ImGui font texture.
+     */
+    static bool ImGuiCreateFontTexture()
+    {
+        // Build texture atlas
+        ImGuiIO& io = ImGui::GetIO();
+        unsigned char* pixels;
+        int width, height;
+
+        // Load as RGBA 32-bits (75% of the memory is wasted, but default font is so small) because
+        // it is more likely to be compatible with user's existing shaders. If your ImTextureId
+        // represent a higher-level concept than just a GL texture id, consider calling
+        // GetTexDataAsAlpha8() instead to save on GPU memory.
+        io.Fonts->GetTexDataAsRGBA32(&pixels, &width, &height);
+
+        // Upload texture to graphics system
+        GLint last_texture;
+        GL_CHECK(glGetIntegerv(GL_TEXTURE_BINDING_2D, &last_texture));
+        GL_CHECK(glGenTextures(1, &g_FontTexture));
+        if (!g_FontTexture) return false;
+        GL_CHECK(glBindTexture(GL_TEXTURE_2D, g_FontTexture));
+        GL_CHECK(glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR));
+        GL_CHECK(glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR));
+        GL_CHECK(glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, pixels));
+
+        // Store our identifier
+        io.Fonts->TexID = (void *)(intptr_t)g_FontTexture;
+
+        // Restore state
+        GL_CHECK(glBindTexture(GL_TEXTURE_2D, last_texture));
+
+        return true;
+    }
+
+    /**
+     * \brief The mouse button callback for GLFW.
+     */
+    static void GlfwMouseButtonCallback(GLFWwindow*, int button, int action, int /*mods*/)
+    {
+        if (action == GLFW_PRESS && button >= 0 && button < 3)
+            g_MousePressed[button] = true;
+    }
+
+    /**
+     * \brief The scroll callback for GLFW.
+     */
+    static void GlfwScrollCallback(GLFWwindow*, double /*xoffset*/, double yoffset)
+    {
+        g_MouseWheel += (float)yoffset; // Use fractional mouse wheel, 1.0 unit 5 lines.
+    }
+
+    /**
+     * \brief The key callback for GLFW.
+     */
+    static void GlfwKeyCallback(GLFWwindow*, int key, int, int action, int mods)
+    {
+        ImGuiIO& io = ImGui::GetIO();
+        if (action == GLFW_PRESS)
+            io.KeysDown[key] = true;
+        if (action == GLFW_RELEASE)
+            io.KeysDown[key] = false;
+
+        (void)mods; // Modifiers are not reliable across systems
```

Comment:
It marks the parameter as unused (to remove a warning), but I just commented out the param name. Leftover from ImGui example.